crates/red_knot/src/salsa_db/semantic/ast_ids.rs

```diff
@@ -0,0 +1,406 @@
+use super::{Db, Jar};
+use crate::ast_ids::NodeKey;
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::visitor::preorder;
+use ruff_python_ast::visitor::preorder::PreorderVisitor;
+use ruff_python_ast::{
+    Expr, ModModule, Stmt, StmtAnnAssign, StmtAssign, StmtClassDef, StmtFunctionDef, StmtImport,
+    StmtImportFrom,
+};
+use ruff_python_parser::Parsed;
+use rustc_hash::FxHashMap;
+use std::collections::VecDeque;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use crate::salsa_db::source::{parse, File};
+
+#[tracing::instrument(level = "debug", skip(db))]
+#[salsa::tracked(jar=Jar, no_eq, return_ref)]
+pub fn ast_ids(db: &dyn Db, file: File) -> AstIds {
+    let parsed = parse(db.upcast(), file);
+
+    AstIds::from_parsed(&parsed)
+}
+
+pub struct AstIds {
+    expressions: IndexVec<ExpressionId, AstNodeRef<Expr>>,
+
+    /// Maps expressions to their expression id. Uses `NodeKey` because it avoids cloning [`Parsed`].
+    expressions_map: FxHashMap<NodeKey, ExpressionId>,
+
+    statements: IndexVec<StatementId, AstNodeRef<Stmt>>,
+
+    statements_map: FxHashMap<NodeKey, StatementId>,
+}
```

Comment:
I haven't, and i wasn't aware of that data structure. I prefer our implementation because we use an `IndexVec` for statements and `expressions` where a lookup is just an array offset whereas `BiMap` would require a hash map lookup.